Initial draft of Python 'Security Policy' (#1804)
python/devguide: security/index.rst

=========================
Security and threat model
=========================

The majority of Python Security Response Team (PSRT)
members are volunteers. Therefore, you must respect this volunteered time
by following this security policy. Repeated failure to
respect the security policy will result in future reports
being rejected or being banned from the `python` GitHub organization, regardless of technical merit.

What types of bugs are vulnerabilities?
---------------------------------------

Not all bugs are vulnerabilities. To avoid causing
duplicate work for PSRT members all potential reports
must be evaluated against the relevant threat models

> prior to being submitted to the PSRT.

Reviewer: Should we ask reporters to cite from the threat model when reporting to ensure that they've done this?

Author: Done in https://github.com/python/devguide/pull/1804/commits/846743420e6b865507ef74f111c511742b0e13ed

The complete threat model for Python and standard library modules
is available in the Threat Model section of the Python Developer Guide.

What versions of Python are accepting reports?
----------------------------------------------

Python accepts security vulnerabilities and will
assign CVE IDs for `supported Python versions`_ that have a status of
`"bugfix" or "security" <python-status>`_. Versions that are not yet
stable (status of `"feature" or "prerelease" <python-status>`_) are not
eligible for CVE IDs. If the vulnerability exclusively exists in
non-stable versions, then the issue should be handled as a public bug issue.

Sometimes features may be marked as
"experimental" in Python, even in a stable Python version.
These features are not eligible for security vulnerabilities.
Instead open a public GitHub issue.

If a vulnerability is platform-dependent, check if the platform is
`supported per :pep:`11`.
Vulnerabilities that exclusively affect unsupported platforms
may not be accepted.

.. _supported Python versions: https://devguide.python.org/versions/
.. _python-status: https://devguide.python.org/versions/#status-key

> What to include and how to structure a vulnerability report?

Reviewer 2: non-blocking, maybe followups: Should we request that people include a practical exploit proof of concept or not?

Also: The recently realized theme around "if it was discovered using a publicly available LLM, we may request that it be worked on in public because we should assume a bunch of others have also discovered it and not reported it - or soon will" might be something to codify in here.  But probably as a followup a bit later?

Reviewer 3: Unfortunately, there are reports that absolutely need a PoC, and others that are just made way more complex when reporters try to create one. So I don't think we can have a clear rule on that one, unfortunately. "Include a PoC if it's not obvious how it may be exploited, and we may request one if you don't but we need it"? I don't like that wording, but it's basically the reality.

On tools, "if it was discovered using a publicly available _tool_, we may request ..." is probably fine. If someone just fires up an open source fuzzer or a new public linter and finds something, it's kinda fair to assume it's not secret. LLMs aren't really _that_ special here, they're just much more versatile than the scanning tools we had before (and more hype, and easier to set up, somehow).

------------------------------------------------------------

For your vulnerability report to be handled efficiently by
the PSRT, the report must include certain information and
be formatted correctly:

* For the initial report and follow-up communications, avoid
  overly long, verbose, or excessive structure (such as headers or tables).
  Ideally reports should be a few sentences describing the vulnerability and
  a proof-of-concept script that reproduces the issue.
* When reporting large numbers or "batches" of vulnerabilities or
  searching for potential vulnerabilities using an LLM, you as a reporter must
  verify the validity of all reports prior to submission to the PSRT.
  PSRT members WILL NOT spend time confirming the validity of reports, only
  whether a valid bug report is a vulnerability or not.
* Do not include severity or CVSS information in your initial report,
  this information will be determined by the PSRT.
* Ideally, include a minimal patch with the mitigation for the report.

> * If the vulnerability only affects certain Python versions, optionally

Reviewer 2: Q: Should we include some text saying the reporter should also attempt to reproduce the problem on the `main` branch and include the result of that?  Ideally having figured out what already merged change fixed the problem?

Author: I think done in https://github.com/python/devguide/pull/1804/commits/846743420e6b865507ef74f111c511742b0e13ed

  include the versions of Python that are affected.
* Reports that do not contain a potential security vulnerability (such as spam
  or requesting compliance or due-diligence work)
  will be discarded without a reply.

How to submit a vulnerability report?
-------------------------------------

Submit all potential security vulnerability reports for CPython

> to GitHub Security Advisories

Reviewer 4: I'd be cautious with this wording, we don't want reports sent to GitHub itself.

=========================
Security and threat model
=========================

The majority of Python Security Response Team
members are volunteers and therefore you must respect this volunteered time
by following this security policy. Repeated failure to
respect the security policy will result in future reports
being rejected, regardless of technical merit.

> 

Author: From @emmatyping: 

> Oh, I see that there is a section about it below. Perhaps moving it up would be better?

Author: Done in https://github.com/python/devguide/pull/1804/commits/846743420e6b865507ef74f111c511742b0e13ed

What types of bugs are vulnerabilities?
---------------------------------------

Not all bugs are vulnerabilities. To avoid causing
duplicate work for PSRT members all potential reports
must be evaluated against the relevant threat model(s)

> prior to being submitted to the PSRT.

Reviewer 3: ```suggestion
Not all bugs are vulnerabilities, and Python is intentionally a highly capable tool
used for a wide variety of tasks in a wide variety of contexts.
Every formally disclosed security vulnerability has an immediate and significant
impact on our users, and so we seriously consider whether it is in our
users' best interests for a bug to be identified as a security vulnerability.

If you believe your issue deserves to potentially cause an emergency response
in thousands of organizations around the world,
please maximize your chances of us agreeing with you by reviewing this threat model
and ensuring that your report meets our criteria for a notifiable vulnerability
(in other words, it gets a CVE ID),
and that your report is clear enough for the PSRT to evaluate it.

To be clear, we consider that a CVE is a notification for our users,
and not a "reward" for reporters. The PSRT is the CNA for Python,
and is the sole decision maker for whether an issue is notifiable.
```

Possibly a bit dramatic, but feel free to tone it down 😉

Reviewer 4: "this threat model" - since it doesn’t exist yet, I think this change should be postponed.

Maybe mark it as `.. important::`?

Reviewer 3: We immediately go on to describe the assumptions we make, which is close enough to being a threat model that I think we can just call it that. For practical purposes, that's all that really matters.

Author: @zooba I hope you are okay if I leave the policy as-is in this section. Leaving this discussion unresolved so give me an indication if this is okay with you.

Reviewer 3: The first two paragraphs I proposed are obviously me venting frustration, but if you can get some aspect of the third paragraph in somewhere early I think that would help set expectations for submitters (I don't think we say anything about bug bounties on this page, but being clear that we don't award those either would be aligned).

Occasionally we get people researching and reporting for the sole purpose of putting CVEs on their resume, and they get snippy if we don't agree that it's a vulnerability or don't agree to their credit terms (or give it to an earlier report). So I'd like to see "the CVE is for our users, not for you" strongly implied somewhere.

Author: We discussed this a bit more, I think we're going to keep this section as-is for now as I believe most bad behavior we're trying to guard against will be covered by other parts of the security policy. We do want to still encourage researchers to examine Python and be rewarded for their efforts through attribution on CVE IDs, but only if their interactions with our security team are appropriate :)

Vulnerabilities must be exploitable from code, configurations,
pre-conditions, and deployments that might feasibly exist in
the real-world. For example, a vulnerability only affecting code
that does not make sense to write in a production program
will not be accepted as a vulnerability.

Documented functionality will not be considered a vulnerability.
For example, ``pickle``, ``marshal``, ``shelve``, ``eval``, and ``exec``
are documented to execute arbitrary Python code that is supplied as data.
The ``ctypes`` module is documented to enable modifying arbitrary locations
in memory.

Vulnerabilities must not depend on malicious control of:

* what Python code is executed by the interpreter
* locations where code is loaded prior to execution (such as current working
  directory, ``PYTHONPATH``)
* configuration files
* environment variables
* command line options
* installed packages or modules
* `.pth files <https://docs.python.org/3/library/site.html>`__
* caches or ``.pyc`` files

Vulnerabilities that affect availability (such as DoS, ReDoS) must be

> triggerable with data inputs that are reasonably sized for the use-case.

Reviewer 4: Maybe: "...Otherwise if that is not the case and the module has an explicit warning they can be reported publicly"

If a vulnerability is platform-dependent, check if the platform is a
`supported platform per PEP 11 <https://peps.python.org/pep-0011/>`__.
Vulnerabilities that exclusively affect unsupported platforms
may not be accepted.

.. _supported Python versions: https://devguide.python.org/versions/
.. _python-status: https://devguide.python.org/versions/#status-key

What to include and how to structure a vulnerability report?
------------------------------------------------------------

For your vulnerability report to be handled efficiently by
the PSRT, the report must include certain information and
be formatted correctly:

* For the initial report and follow-up communications, avoid
  overly long, verbose, or excessive structure (such as headers or tables).
  Ideally reports should be a few sentences describing the vulnerability and
  a proof-of-concept script that reproduces the issue.
* When reporting large numbers or "batches" of vulnerabilities or
  searching for potential vulnerabilities using an LLM, you as a reporter must
  verify the validity of all reports prior to submission to the PSRT.
  PSRT members WILL NOT spend time confirming the validity of reports, only

>   whether a valid bug report is a vulnerability or not.

Reviewer 5: I think this is maybe confusingly worded. We probably should verify a vulnerability is a vulnerability.

* Do not include severity or CVSS information in your initial report,
  this information will be determined by the PSRT.
* Optionally, include a minimal patch with the mitigation for the report.
* If the vulnerability only affects certain Python versions, optionally

>   include the versions of Python that are affected.

Reviewer 3: ```suggestion
* Always include the versions of Python that were tested, 
  and indicate which were found to be vulnerable.
```

* Reports that do not contain a potential security vulnerability (such as spam
  or requesting compliance or due-diligence work)
  will be discarded without a reply.

> 

Reviewer 3: ``` 
* Submit as plain-text only, including attachments.
  No PDFs, binaries, or other files that cannot be safely reviewed.
  If your proof-of-concept depends on a specially constructed binary file,
  please include a script to construct it rather than the file itself.
```

Reviewer 4: I suggest adding "notebooks," we always get those :-/

* Once a public pull request containing a fix is merged to CPython,
  the advisory and CVE record will be published with attribution.

For more information about how the PSRT handles vulnerabilities,
`consult the Python Developer Guide <https://devguide.python.org/developer-workflow/psrt/#triaging-a-vulnerability-report>`__.

PSF Code of Conduct
-------------------

Well-being and safety of the Python Security Response Team members is
prioritized over the technical merit of vulnerability reports.
Despite communications being private, vulnerability reporting is subject
to the `PSF Code of Conduct`_. Violations will be reported to the Code of Conduct
team with undisclosed vulnerability information removed, if applicable.

.. _GHSA: https://github.com/python/cpython/security/advisories/new
.. _PSF Code of Conduct: https://policies.python.org/python.org/code-of-conduct/

CVE Numbering Authority (CNA)
-----------------------------

The Python and pip projects are scoped under the
`Python Software Foundation CVE Numbering Authority <CNA>`__
(CNA). This means you must submit all security
vulnerability reports to the PSRT to receive

> a CVE ID for Python or pip. To reach the PSF

Reviewer 3: ```suggestion
vulnerability reports to the PSRT for a CVE ID
to be issued for Python or pip. To reach the PSF
```
